ships/0005-local-source-upload.md

```diff
@@ -0,0 +1,262 @@
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [docs](/docs/)
+
+## Summary
+
+Proposal to improve developer-experience by allowing to upload a local repository clone to the
+Kubernetes cluster, and use Shipwright Build Controller to produce a container-image out of it.
+
+As a developer, I can try Shipwright before opening a pull-request, and take advantage of the cluster
+resources to create the image. Effectively moving shipwright closer to the developer's inner-loop, in
+other words closer to the frequent activities while writing code ([1][InnerLoopA], [2][InnerLoopB]),
+in other words we are "moving left before _shipping-right_".
+
+## Motivation
+
+Shipwright Build-Controller allows teams to build container-images out of Git repositories, register
+and track their application builds in a Kubernetes native environment. This enhancement proposal
+extends this ability for developers to upload uncommitted changes and use the same machinery during
+the inner-loop as well.
+
+As a developer you can try out changes before opening a pull-request, make sure the final
+container-image will work out as expected, and effectively improve the developer experience.
+
+The approach proposed here is complementary to existing [EP][LocalSourceCodeEP] detailing local
+source upload to a container-registry and later on self-extracting the image before the actual
+container build process starts.
+
+### Goals
+
+- Improve developer experience, local clone can produce a container-image using Shipwright machinery
+- Move Shipwright CLI closer to developer's inner-loop
+- Complementary to [upload-to-container-registry (EP)][LocalSourceCodeEP], uploading data directly to
+the cluster instead
+
+## Proposal
+
+This proposal includes changes in Shipwright components, and adds a new helper application to aid
+data upload process.
+
+### Sequence Diagram
+
+The following sequence-diagram shows the relationship between the components and explains the order
+of execution. In a nutshell, the CLI creates a `BuildRun` object, with settings to wait for data
+upload, which can only start when the Build-Controller instructs Kubernetes to create this new pod.
+The new pod, on the other hand, needs to be running and ready to receive external data.
+
+With the data upload complete, the build-pod can continue to execute the build-strategy chosen,
+producing a container-image at the end.
+
+![source-upload sequence diagram](assets/0005-source-upload-sequence-diagram.png)
+
+To explain the sequence of events in more detail, please consider the following topics.
+
+#### Prepare _(Step 1 to 3)_
+
+The preparation consists on declaring a BuildRun object _(1)_ with configuration for the
+Build-Controller, to instead of appending the regular "`git clone`" step, to wait for the user input
```

Comment:
```suggestion
Build-Controller, to instead of appending the regular "`git clone`" step at runtime, it will block ( _wait_ ) the execution of the strategy steps until the user input is available in the pod.
```